### PR TITLE
8255968: Confusing error message for inaccessible constructor

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

When using inaccessible constructor, compiler would output some confusing error message occasionally. These confusing error message is related to the constructor order. See [JDK-8255968](https://bugs.openjdk.java.net/browse/JDK-8255968) for more information.
This patch solves this bug, regardless of the construction order, compiler can always output expected message.
Thank you for taking the time to review.

Best Regards.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ❌ (6/6 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test tasks**
- [Linux x64 (build debug)](https://github.com/lgxbslgx/jdk/runs/1469765121)
- [Linux x64 (build hotspot minimal)](https://github.com/lgxbslgx/jdk/runs/1469765132)
- [Linux x64 (build hotspot no-pch)](https://github.com/lgxbslgx/jdk/runs/1469765125)
- [Linux x64 (build hotspot optimized)](https://github.com/lgxbslgx/jdk/runs/1469765133)
- [Linux x64 (build hotspot zero)](https://github.com/lgxbslgx/jdk/runs/1469765127)
- [Linux x64 (build release)](https://github.com/lgxbslgx/jdk/runs/1469765109)

### Issue
 * [JDK-8255968](https://bugs.openjdk.java.net/browse/JDK-8255968): Confusing error message for inaccessible constructor

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1389/head:pull/1389`
`$ git checkout pull/1389`

### CODE REVIEW

**Review by mcimadamore — COMMENTED** [review]
While I somewhat sympathize with the problem you are trying to solve, note that, by doing this, the output of the compiler will regress in case where not _all_ the overload candidates are not accessible. E.g.

```
m(1): //client

m();
private m(int);
```

In this case I guess it's less clear as what "should" happen: does the user want to call the first, or second version of `m` ? Both cases are possible - e.g. the user might accidentally call the private version of some method (in which the compiler is right in pointing out that the callsite is wrong), or the user indeed might want to access the private method (in which case your patch to shift the focus onto the access error might seem more appropriate).

I'm a bit skeptical that there exist a silver bullet for these kind of mixed cases. Technically speaking, from a JLS perspective, non-accessible methods should not even take place in overload resolution - so the current compiler behavior is cases where both accessible and inaccessible methods are present seems more appropriate.

In other words, if we really want to improve the status quo, we should detect cases where _all_ candidates are not accessible. Or, we should question whether javac's behavior of tagging inaccessible methods with the kind `WRONG_MTH`/`WRONG_MTHS` is correct in the first place.

It would be nice, I guess, if we could make the inaccessible error more powerful, so that we could keep adding inaccessble symbols as we go along, so that javac could display something like "cannot resolve symbol `m` - but some inaccessible symbols have been found <list>".

**Comment by lgxbslgx** [comment]
Currently, the compiler output the error message depending on the order of the methods. Please see the below test cases.

- the `private Test(int x) {}` is at the end of other(one or more) method(s).
```
class T8255968 {
    Test c = new Test(0);
}
class Test {
    private Test(String[] x) {}
    Test(String x) {}
    private Test(int x) {}  // at the end
}

current output:
error: Test(int) has private access in Test
    Test c = new Test(0);
             ^
1 error
```
- the `private Test(int x) {}` is not at the end of other(only one) method.
```
class T8255968 {
    Test c = new Test(0);
}
class Test {
    private Test(int x) {}  // not at the end
    Test(String x) {} // can be private, output same error message
}

current output:
error: incompatible types: int cannot be converted to String
    Test c = new Test(0);
                      ^
Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output
1 error
```

- the `private Test(int x) {}` is not at the end of other(more than one) methods.
```
class T8255968 {
    Test c = new Test(0);
}
class Test {
    private Test(String x) {}
    private Test(int x) {}  // not at the end
    Test(String[] x) {}
}

current output:
error: no suitable constructor found for Test(int)
    Test c = new Test(0);
             ^
    constructor Test.Test(String) is not applicable
      (argument mismatch; int cannot be converted to String)
    constructor Test.Test(String[]) is not applicable
      (argument mismatch; int cannot be converted to String[])
1 error
```
These output messages which depending on the order of the methods would confuse users. Because the order of the methods  should not have any influence.

@mcimadamore According to your comment, I want to unify the output messages as below.

- the `private Test(int x) {}` is at the end of other(one or more) method(s).
```
class T8255968 {
    Test c = new Test(0);
}
class Test {
    private Test(String[] x) {}
    Test(String x) {}
    private Test(int x) {}  // at the end
}

expected output:
error: no suitable constructor found for Test(int)
    Test c = new Test(0);
             ^
    constructor Test.Test(String[]) is not applicable
      (argument mismatch; int cannot be converted to String[])
    constructor Test.Test(int) is not applicable
      (Test(int) has private access in Test)
    constructor Test.Test(String) is not applicable
      (argument mismatch; int cannot be converted to String)
1 error
```
- the `private Test(int x) {}` is not at the end of other(only one) method.
```
class T8255968 {
    Test c = new Test(0);
}
class Test {
    private Test(int x) {}  // not at the end
    Test(String x) {} // can be private, output same error message
}

expected output:
error: no suitable constructor found for Test(int)
    Test c = new Test(0);
             ^
    constructor Test.Test(int) is not applicable
      (Test(int) has private access in Test)
    constructor Test.Test(String) is not applicable
      (argument mismatch; int cannot be converted to String)
1 error
```

- the `private Test(int x) {}` is not at the end of other(more than one) methods.
```
class T8255968 {
    Test c = new Test(0);
}
class Test {
    private Test(String x) {}
    private Test(int x) {}  // not at the end
    Test(String[] x) {}
}

expected output:
error: no suitable constructor found for Test(int)
    Test c = new Test(0);
             ^
    constructor Test.Test(String) is not applicable
      (argument mismatch; int cannot be converted to String)
    constructor Test.Test(int) is not applicable
      (Test(int) has private access in Test)
    constructor Test.Test(String[]) is not applicable
      (argument mismatch; int cannot be converted to String[])
1 error
```

And the following two situations are not modified.
```
class T8255968 {
    Test c = new Test(0);
}
class Test {
    private Test(int x) {}
}

current and expected output:
error: Test(int) has private access in Test
    Test c = new Test(0);
             ^
1 error
```
```
class T8255968 {
    Test c = new Test(0);
}
class Test {
    Test(String x) {}
}

current and expected output:
error: incompatible types: int cannot be converted to String
    Test c = new Test(0);
                      ^
Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output
1 error
```

I implement this unified format and test the feature locally. @mcimadamore If you agree with these unified output messages, I would update my code for reviewing.

**Comment by mcimadamore** [comment]
> 
> I implement this unified format and test the feature locally. @mcimadamore If you agree with these unified output messages, I would update my code for reviewing.

This seems a good directions - but there are few dragons in there. As I said in my previous message, inaccessible members are just not overload resolution candidates and are "skipped" over by the overload selection logic. This is visible in examples like these:

```
class Sup {
   private void m();
}

class Outer {
     void m(String s) { ... }
     void m(Integer s) { ... }

     class Inner extends Sup {
          void test() {
               m();
           }
      }
}
```
In this example, overload selection should look for `m` candidates inside the `Outer` class. If, with a patch, you make access errors more similar to `WRONG_MTH`, or `WRONG_MTHS` kinds (by unifying them) chances are that overload selection will start thinking that it found a candidate in `Sup` and stop there - which would be against the JLS. I'm not saying it cannot be done, but more that it has to be done in a careful way - that is, you can only unify the error messages in situation where you know there's a mix of inaccessible and inapplicable methods, so you know already that overload selection will stop at that class.

Another aspect to look out for/test is method references - whatever we say here for plain method calls, has to hold for method reference selection - so you'll want to create specific tests which aims at showing that whatever change is applied, it also improves the situation with method references (or at least does not cause unwanted regressions there).

**Comment by lgxbslgx** [comment]
I update the code which unifies the error messages. I use the following command to run test locally and all the test passed.
```
make test TEST="jtreg:langtools/tools/javac" CONF=linux-x86_64-server-release JOBS=8

output:
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/langtools/tools/javac                   3503  3503     0     0   
==============================
TEST SUCCESS

Finished building target 'test' in configuration 'linux-x86_64-server-release'
```

Thank you for taking the time to review.

Best Regards.

**Review by mcimadamore — COMMENTED** [review]
Looks a nice generalization. However I still have some concerns:

* the logic, as I feared, is turning `ABSENT` into `WRONG_MTHS` - note that AccessError and InapplicableSymbol/s have different properties when it comes to method lookup - more specifically look at ResolveError::exists which return `false` in one case and `true` in the other. This is used, among other things, to shortcircuit method lookups (see `Resolve::findFun`) - so flipping from `AccessError` to `InapplicableSymbols` is almost guaranteed to cause a change in behavior in the compiler. That said, given the nature of the code involved, is not super trivial to come up with an example which demonstrates the difference - but I'll keep thinking.

* While the tests you add are good, I don't see the corresponding tests for the method reference case.

**Comment by lgxbslgx** [comment]
>  That said, given the nature of the code involved, is not super trivial to come up with an example which demonstrates the difference

I can't find a example to demonstrate the differences too.

> While the tests you add are good, I don't see the corresponding tests for the method reference case.

I add corresponding tests about the method reference cases.

> If an access error shows up and we already have an access error, you add that to the list of "candidates", and return a WRONG_MTHS kind.

It should be:  If an **InapplicableMethodException** occurs and we already have an access error

I add some comments to state the situations and improve readability.

Thank you for taking the time to review.

File: test/langtools/tools/javac/T8255968/T8255968_10.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+class T8255968 {
```
Comment (mcimadamore) [inline test/langtools/tools/javac/T8255968/T8255968_10.java:24]: Can you please disambiguate all these class names? Having the same class name in multiple tests can create havocs, especially when running in concurrent mode. Please make sure that classes have different names, and that the auxiliary classes defined in these test are, instead, defined as nested classes (so that, again, everything will be encapsulated on a test-by-test basis, w/o any risk of a test picking up a class from another test).

File: test/langtools/tools/javac/T8255968/T8255968.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8255968
+ * @summary Confusing error message for inaccessible constructor
```
Comment (mcimadamore) [inline test/langtools/tools/javac/T8255968/T8255968.java:27]: Any reason as to why you went for this single file which lists all tests rather than adding jtreg tags in all the various test cases? AFAIK, in the javac code base the latter style is often preferred.

**Review by mcimadamore — COMMENTED** [review]
I left some additional comments on tests. I plan to do some experiments to validate that we are not going to introduce bad regressions.

File: test/langtools/tools/javac/T8255968/T8255968_10.java (hunk 1 above)
Comment (lgxbslgx) [reply test/langtools/tools/javac/T8255968/T8255968_10.java:24]: It sounds a great way to avoid some problem. I would revise my code to meet this style. Thanks for the suggestion.

File: test/langtools/tools/javac/T8255968/T8255968.java (hunk 2 above)
Comment (lgxbslgx) [reply test/langtools/tools/javac/T8255968/T8255968.java:27]: I have seem these two styles in the test code base. I don't prefer any style because these two styles both work well. It is better to have a standard style to guide developers. But now it has no such standard.

> AFAIK, in the javac code base the latter style is often preferred.

According to your comment, I would revise the tests to meets the latter style.

**Comment by lgxbslgx** [comment]
The old style which is shown below is not good.
```
class T8255968 {
    Test c = new Test(0);
}

class Test {
    private Test(int x) {}
}
```

> and that the auxiliary classes defined in these test are, instead, defined as nested classes

I change it to the following style according to the comment. But the compiler compiles it successfully because the outer class can visit private method of its inner class.
```
class T8255968_1 {
    Test c = new Test(0);

    class Test {
        private Test(int x) {}
    }
}

```

So I would like to use the following style(don't use nested classes).
```
class T8255968_1 {
    T8255968Test1 c = new T8255968Test1(0);
}

class T8255968Test1 {
    private T8255968Test1(int x) {}
}
```

@mcimadamore Do you agree with this style?

**Comment by mcimadamore** [comment]
> @mcimadamore Do you agree with this style?

Yeah - of course nesting things is gonna mess up with access checks - sorry for sending you down the wrong direction!

You could prefix every class in the test with `T8255968_1` e.g.

```
class T8255968_1 {
    T8255968Test1 c = new T8255968_1_Test(0);
}

class T8255968_1_Test {
    private T8255968_1_Test(int x) {}
}
```

**Comment by lgxbslgx** [comment]
I update the code about test cases. Thank you for taking the time to review.

**Comment by mcimadamore** [comment]
I haven't forgotten about this - we're running some extensive corpus analysis to make sure that the patch won't introduce any unwanted regressions. Please bear with us :-)

**Comment by mcimadamore** [comment]
All the internal testing went ok. At this point I see no reason to delay this further. Thanks.

**Review by mcimadamore — APPROVED** [review]
Looks good

**Comment by lgxbslgx** [comment]
Thank you for the review.

@mcimadamore 
Could you please sponsor this patch?
Thanks again.
